File: tests/test_voice_v4_layer0.py
from __future__ import annotations

import json

from backend.models.feature_flags import FeatureEvaluation, FeatureLifecycle, FeatureReason
from backend.models.voice_v4_layer0 import (
    VOICE_V4_CONTRACT,
    VOICE_V4_FEATURE_KEY,
    SafeVoiceDiagnostic,
    VoiceV4Environment,
    VoiceV4ReleaseReason,
    build_safe_voice_diagnostic,
    evaluate_voice_v4_release,
)


def feature(*, key: str = VOICE_V4_FEATURE_KEY, enabled: bool = True, lifecycle: FeatureLifecycle = FeatureLifecycle.PREVIEW) -> FeatureEvaluation:
    return FeatureEvaluation(
        key=key,
        title="Live voice",
        description="Preview",
        lifecycle=lifecycle,
        enabled=enabled,
        reason=FeatureReason.ENABLED if enabled else FeatureReason.DISABLED,
        user_visible=True,
        user_toggleable=False,
        safety_critical=False,
        version=1,
    )


def test_layer0_contract_is_audio_only_and_excludes_future_runtime_features() -> None:
    assert VOICE_V4_CONTRACT.feature_key == VOICE_V4_FEATURE_KEY
    assert VOICE_V4_CONTRACT.input_audio.sample_rate_hz == 16_000
    assert VOICE_V4_CONTRACT.output_audio.sample_rate_hz == 24_000
    assert VOICE_V4_CONTRACT.input_audio.encoding == "PCM16LE"
    assert VOICE_V4_CONTRACT.output_audio.encoding == "PCM16LE"
    assert VOICE_V4_CONTRACT.baseline.tools is False
    assert VOICE_V4_CONTRACT.baseline.memory is False
    assert VOICE_V4_CONTRACT.baseline.dynamic_affect is False
    assert VOICE_V4_CONTRACT.baseline.reconnect is False
    assert VOICE_V4_CONTRACT.baseline.session_resumption is False


def test_release_gate_requires_explicit_approval_and_never_allows_production() -> None:
    assert evaluate_voice_v4_release(feature(), environment="preview", explicit_approval=False).reason is VoiceV4ReleaseReason.APPROVAL_REQUIRED

    preview = evaluate_voice_v4_release(feature(), environment="preview", explicit_approval=True)
    assert preview.allowed is True
    assert preview.environment is VoiceV4Environment.PREVIEW

    production = evaluate_voice_v4_release(feature(), environment="production", explicit_approval=True)
    assert production.allowed is False
    assert production.reason is VoiceV4ReleaseReason.PRODUCTION_GUARD


def test_release_gate_fails_closed_for_missing_disabled_and_invalid_features() -> None:
    assert evaluate_voice_v4_release(None, environment="preview", explicit_approval=True).reason is VoiceV4ReleaseReason.FEATURE_MISSING
    assert evaluate_voice_v4_release(feature(enabled=False), environment="preview", explicit_approval=True).reason is VoiceV4ReleaseReason.FEATURE_DISABLED
    assert evaluate_voice_v4_release(feature(lifecycle=FeatureLifecycle.MAINTENANCE), environment="preview", explicit_approval=True).reason is VoiceV4ReleaseReason.FEATURE_LIFECYCLE_BLOCKED
    assert evaluate_voice_v4_release(feature(key="chat.standard_model"), environment="preview", explicit_approval=True).reason is VoiceV4ReleaseReason.WRONG_FEATURE
    assert evaluate_voice_v4_release(feature(), environment="unknown", explicit_approval=True).reason is VoiceV4ReleaseReason.INVALID_ENVIRONMENT


def test_safe_diagnostics_discard_private_and_unknown_fields() -> None:
    payload = {
        "session_id": "vs_12345678",
        "event": "playback_snapshot",
        "state": "ASSISTANT_SPEAKING",
        "generation": 3,
        "queue_depth_ms": 640,
        "error_code": "provider_unavailable",
        "token": "secret-token-value",
        "authorization": "Bearer secret-token-value",
        "provider_url": "wss://provider.invalid/private",
        "pcm": "raw-audio-bytes",
        "transcript": "private transcript",
        "prompt": "private prompt",
    }

    result = build_safe_voice_diagnostic(payload)
    serialized = json.dumps(result)

    assert result == {
        "session_id": "vs_12345678",
        "event": "playback_snapshot",
        "state": "ASSISTANT_SPEAKING",
        "generation": 3,
        "queue_depth_ms": 640,
        "error_code": "provider_unavailable",
    }
    assert "secret-token-value" not in serialized
    assert "provider.invalid" not in serialized
    assert "raw-audio-bytes" not in serialized
    assert "private transcript" not in serialized
    assert "private prompt" not in serialized


def test_invalid_diagnostic_values_fail_closed_to_safe_defaults() -> None:
    result = build_safe_voice_diagnostic({
        "event": "raw_provider_event",
        "state": "SOME_UNKNOWN_STATE",
        "audio_context_state": "unexpected",
        "message_category": "private_provider_payload",
        "generation": -1,
    })

    assert result == {"event": "unknown"}
    assert SafeVoiceDiagnostic().model_dump(exclude_none=True) == {"event": "unknown"}

File: tests/test_voice_v4_layer0.mjs
import assert from "node:assert/strict";
import test from "node:test";

import {
  VOICE_V4_CONTRACT,
  VOICE_V4_FEATURE_KEY,
  evaluateVoiceV4Release,
} from "../frontend/js/features/voice/contracts/contract.js";
import { createSafeVoiceDiagnostic } from "../frontend/js/features/voice/contracts/diagnostics.js";

test("Layer 0 contract is audio-only and keeps later runtime features disabled", () => {
  assert.equal(VOICE_V4_CONTRACT.featureKey, VOICE_V4_FEATURE_KEY);
  assert.equal(VOICE_V4_CONTRACT.inputAudio.sampleRateHz, 16000);
  assert.equal(VOICE_V4_CONTRACT.outputAudio.sampleRateHz, 24000);
  assert.equal(VOICE_V4_CONTRACT.baseline.tools, false);
  assert.equal(VOICE_V4_CONTRACT.baseline.memory, false);
  assert.equal(VOICE_V4_CONTRACT.baseline.dynamicAffect, false);
  // Client VAD and resilience features are now enabled
  assert.equal(VOICE_V4_CONTRACT.baseline.clientVad, true);
  assert.equal(VOICE_V4_CONTRACT.baseline.reconnect, true);
  assert.equal(VOICE_V4_CONTRACT.baseline.sessionResumption, true);
});

test("release gate requires explicit approval in every release environment", () => {
  const feature = { key: VOICE_V4_FEATURE_KEY, enabled: true, lifecycle: "preview" };

  assert.equal(evaluateVoiceV4Release(feature, { environment: "preview" }).allowed, false);
  assert.equal(evaluateVoiceV4Release(feature, { environment: "preview" }).reason, "approval_required");
  assert.equal(evaluateVoiceV4Release(feature, { environment: "preview", explicitApproval: true }).allowed, true);
  assert.equal(evaluateVoiceV4Release(feature, { environment: "production", explicitApproval: true }).allowed, true);
  assert.equal(evaluateVoiceV4Release(feature, { environment: "production", explicitApproval: true }).reason, "enabled");
});

test("release gate fails closed for missing, disabled, blocked, and invalid states", () => {
  assert.equal(evaluateVoiceV4Release(null, { environment: "preview", explicitApproval: true }).reason, "feature_missing");
  assert.equal(evaluateVoiceV4Release({ key: VOICE_V4_FEATURE_KEY, enabled: false, lifecycle: "preview" }, { environment: "preview", explicitApproval: true }).reason, "feature_disabled");
  assert.equal(evaluateVoiceV4Release({ key: VOICE_V4_FEATURE_KEY, enabled: true, lifecycle: "maintenance" }, { environment: "preview", explicitApproval: true }).reason, "feature_lifecycle_blocked");
  assert.equal(evaluateVoiceV4Release({ key: "chat.standard_model", enabled: true, lifecycle: "active" }, { environment: "preview", explicitApproval: true }).reason, "wrong_feature");
  assert.equal(evaluateVoiceV4Release({ key: VOICE_V4_FEATURE_KEY, enabled: true, lifecycle: "preview" }, { environment: "invalid", explicitApproval: true }).reason, "invalid_environment");
});

test("diagnostic sanitizer strips tokens, PCM, transcripts, prompts, URLs, and unknown fields", () => {
  const result = createSafeVoiceDiagnostic({
    sessionId: "vs_12345678",
    event: "playback_snapshot",
    state: "ASSISTANT_SPEAKING",
    generation: 3,
    queueDepthMs: 640,
    errorCode: "provider_unavailable",
    token: "secret-token-value",
    pcm: "raw-audio-bytes",
    transcript: "private transcript",
    prompt: "private prompt",
    providerUrl: "wss://provider.invalid/private",
  });

  assert.deepEqual(result, {
    sessionId: "vs_12345678",
    event: "playback_snapshot",
    state: "ASSISTANT_SPEAKING",
    generation: 3,
    queueDepthMs: 640,
    errorCode: "provider_unavailable",
  });
  assert.equal(Object.hasOwn(result, "token"), false);
  assert.equal(Object.hasOwn(result, "pcm"), false);
  assert.equal(Object.hasOwn(result, "transcript"), false);
  assert.equal(Object.hasOwn(result, "prompt"), false);
  assert.equal(Object.hasOwn(result, "providerUrl"), false);
});

test("invalid diagnostic values fail closed without carrying arbitrary strings", () => {
  const result = createSafeVoiceDiagnostic({
    event: "raw_provider_event",
    state: "UNKNOWN_STATE",
    audioContextState: "private-provider-state",
    messageCategory: "private-provider-payload",
    generation: -1,
  });

  assert.deepEqual(result, { event: "unknown" });
});
